src/wallet/wallet.cpp

```diff
@@ -1574,14 +1574,14 @@ CWallet::ScanResult CWallet::ScanForWalletTransactions(const uint256& start_bloc
 {
     int64_t nNow = GetTime();
     int64_t start_time = GetTimeMillis();
-
     assert(reserver.isReserved());
-
-    uint256 block_hash = start_block;
     ScanResult result;
 
     WalletLogPrintf("Rescan started from block %s...\n", start_block.ToString());
 
+    uint256 block_hash = start_block;
+    const auto filter_set = GetLegacyScriptPubKeyMan()->GetAllRelevantScriptPubKeys();
```

Comment:
Setting the filter_set outside the rescan loop means that this doesn't get updated when the keypool is topped up. That means that any payments to an address beyond the first 1000 will be missed on rescan.